accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,165 @@
+
+- ASP.NET - also building shared framework source generators.
+- .NET Libraries team - building multiple shared framework source generators.
+- Mono team - interested in providing source generators.
+- .NET Runtime team - interested in providing source generators.
+- Roslyn team - responsible for source generator infrastructure.
+- .NET SDK team - responsible for resolving shared frameworks and doing conflict resolution with content.
+- NuGet team - FYI only, no changes requested
+
+## Design
+
+### Framework provided source generators
+
+Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not be used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the source generators specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the source generators specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.
+
+Source generators, analyzers, and code fixes are all represented and communicated in the same manner to the compiler, as analyzer dlls using the `-analyzer` switch.  Though this document is motivated by describing the process for source generators it will work in the same manner for analyzers and in some places we may refer to the assemblies containing source generators as analyzers due to the way they are understood by the compiler, package manager, and SDK.
+
+To facilitate this, we will package source generators inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [NuGet conventions for describing analyzers](https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions) with respect to the location in the package, however the SDK will not probe this path.  All source generators and their metadata will be listed in FrameworkList.xml, similar to references.  Analyzers will be added as `File` elements with `Type="Analyzer"` (constrasting to `Type="Managed"` which is currently used for references) and optionally `Language="cs|vb"`.  Omitting the `Language` attribute means the analyzer applies to all languages.  The .NET SDK will be responsible for locating the appropriate source generators from the refpack, based on the project's `$(Language)` and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other `@(Analyzer)` sources as described in the following section.  When viewed from the IDE it should be clear that the source of the `Analyzer` is the framework reference, similarly to how it's clear that the source of NuGet package analyzers come from packaages.
```

Comment:
@dsplaisted please review.  I chose `<File Type="Analyzer"` but could also imagine us using `<Analyzer` if we were worried that older SDK's might ever see these files and mistreat analyzers as references.

Reply:
Looks good.  We'll have to figure out with the project system team how we'll annotate the Analyzer items so that they can associate them with the right framework reference in solution explorer.

Reply:
I wrote this based on what I would expect without actually checking how things work today.  It's entirely possible that this doesn't work well for package-analyzers today.  I did notice that the IDE has a separate node for analyzers and lists the rules that they expose.